src/arithmetic/algebraic_expression.c

```diff
@@ -203,41 +203,81 @@ AlgebraicExpressionResult *AlgebraicExpression_Execute(AlgebraicExpression *ae)
             assert(GrB_transpose(transposed, NULL, NULL, A, NULL) == GrB_SUCCESS);
             C = transposed;
         }
-    }
-
-    if(operand_count > 1) {
+    } else {
         GrB_Descriptor desc;
         GrB_Descriptor_new(&desc);
-        if(ae->operands[0].transpose) GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
 
         GrB_Index nrows;
         GrB_Matrix_nrows(&nrows, A);
         GrB_Matrix_new(&C, GrB_BOOL, nrows, nrows);
-        
-        for(int i = 1; i < operand_count; i++) {
-            AlgebraicExpressionOperand term = ae->operands[i];
-            GrB_Matrix B = term.operand;
-            if(term.transpose) GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
-
-            _AlgebraicExpression_Execute_MUL(C, A, B, desc);
-            A = C;
 
-            // Graph matrices mustn't be modified, restore to original form.
-            // TODO: Verify that descriptor indeed changes underline input.
-            // if(term.transpose) GrB_transpose(B, NULL, NULL, B, NULL);
+        AlgebraicExpressionOperand leftTerm;
+        AlgebraicExpressionOperand rightTerm;
+
+        while(ae->operand_count > 1) {
+            // Pick matrix with min NNZ.
+            int minMatPos = _AlgebraicExpression_MinMatPos(ae);
+            int secondOperandPos = -1;
+            // Pick second operand, either left or right of minMatPos.
+            if(minMatPos == ae->operand_count-1) {
+                // Last operand, multiply to the right.
+                secondOperandPos = minMatPos-1;
+                rightTerm = ae->operands[minMatPos];
+                leftTerm = ae->operands[secondOperandPos];
+            } else if(minMatPos == 0) {
+                // First operand, multiply to the left.
+                secondOperandPos = 1;
+                leftTerm = ae->operands[0];
+                rightTerm = ae->operands[secondOperandPos];
+            } else {
+                /* Intermidate operand, prefer second operand
+                * to have the least NNZ between the two options. */
```

Comment:
I can't entirely visualize what factors should be the strongest factors for choosing the second operand here.

It might be worth favoring a term if it matches the `transpose` flag of the `minMatPos` term, perhaps?

Reply:
Why do you think transpose has is relevant? 

I've a done a quick test and it seems like it's better to have the matrix with the minimum number of NNZ multiplied to the right, I've yet to understand why,  but here are the result of my test:

S-stands for sparse
D-stands for dense.
SD Sparse times Dense
DS Dense times Sparse

Performing 10 comparison of 10000 by 10000 matrices
Creating matrix with 500000 nvals (sparse)
Creating matrix with 10000000 nvals (dense)
sdSum 96469 - overall time performing S*D
dsSum 70630  - overall time performing D*S
sdMax 11148 - max time performing a single S*D
dsMax 7165 - max time performing a single D*S
sdMin 9106  - min time performing a single S*D
dsMin 6950 - min time performing a single D*S
sdAvg 9646.000000 - avg time performing a single S*D
dsAvg 7063.000000  - avg time performing a single D*S

I'll check with Tim, in the meanwhile I'll change my logic to prefer DS.

Reply:
I suppose that transpose conditions can't help here, actually. My thinking was that we might be able to remove some operations when multiplying two adjacent to-be-transposed matrices because `A'B' == (BA)'`, but I don't think we can assume that our matrices can be commuted, so that's not actually helpful.

Reply:
That's an interesting thought, I think it would work, although commuting the matrices in some cases has harsh consequences